Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

```diff
@@ -71,26 +94,60 @@ variable {R : Type*} [CommSemiring R]
 
 variable (m) in
 /-- the degree of a multivariate polynomial with respect to a monomial ordering -/
-def degree {R : Type*} [CommSemiring R] (f : MvPolynomial σ R) : σ →₀ ℕ :=
+def degree (f : MvPolynomial σ R) : σ →₀ ℕ :=
   m.toSyn.symm (f.support.sup m.toSyn)
 
 variable (m) in
 /-- the leading coefficient of a multivariate polynomial with respect to a monomial ordering -/
-def leadingCoeff {R : Type*} [CommSemiring R] (f : MvPolynomial σ R) : R :=
+def leadingCoeff (f : MvPolynomial σ R) : R :=
   f.coeff (m.degree f)
 
 @[deprecated (since := "2025-01-31")] alias lCoeff := leadingCoeff
 
+variable (m) in
+/-- A multivariate polynomial is `Monic` with respect to a monomial order
+if its leading coefficient (for that monomial order) is 1. -/
+def Monic (f : MvPolynomial σ R) : Prop :=
+  m.leadingCoeff f = 1
+
+@[nontriviality] theorem Monic.of_subsingleton [Subsingleton R] {f : MvPolynomial σ R} :
+    m.Monic f :=
+  Subsingleton.eq_one (m.leadingCoeff f)
+
+instance Monic.decidable [DecidableEq R] (f : MvPolynomial σ R) :
+    Decidable (m.Monic f) := by
+  unfold Monic; infer_instance
+
+@[simp]
+theorem Monic.leadingCoeff_eq_one {f : MvPolynomial σ R} (hf : m.Monic f) : m.leadingCoeff f = 1 :=
+  hf
+
+theorem Monic.coeff_degree {f : MvPolynomial σ R} (hf : m.Monic f) : f.coeff (m.degree f) = 1 :=
+  hf
+
 @[simp]
 theorem degree_zero : m.degree (0 : MvPolynomial σ R) = 0 := by
   simp [degree]
 
+@[simp, nontriviality] theorem degree_subsingleton [Subsingleton R] {f : MvPolynomial σ R} :
+    m.degree f = 0 := by
+  rw [Subsingleton.eq_zero f, degree_zero]
```

Comment:
I'm still unsure this lemma is worth having. We don't have `add_subsingleton [AddGroup G] [Subsingleton G] (a b : G) : a + b = a` etc

Reply:
The point is the `nontriviality` tag. Without that lemma (or some other), the tactic doesn't conclude.

Reply:
Can you complain on Zulip that you need to write this lemma for `nontriviality` to work?

Reply:
I think this `nontriviality` lemma is reasonable.

> I'm still unsure this lemma is worth having. We don't have `add_subsingleton [AddGroup G] [Subsingleton G] (a b : G) : a + b = a` etc

Because in this case you'd better have this `@[simp, nontriviality]` lemma instead: `eq_zero [Zero G] [Subsingleton G] (a : G) : a = 0`

Reply:
> Because in this case you'd better have this `@[simp, nontriviality]` lemma instead: `eq_zero [Zero G] [Subsingleton G] (a : G) : a = 0`

This lemma would significantly slow simp down as it has a very general discrimination key

Reply:
OK, maybe only `@[nontriviality]`